File: migration_project/ui_qt/docks/project_dock.py
"""Project and resource tree dock with grouped resources."""

from __future__ import annotations

from pathlib import Path
from typing import Iterable

from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import (
    QMenu,
    QTreeWidget,
    QTreeWidgetItem,
    QVBoxLayout,
    QWidget,
)

from ..i18n import tr


class ProjectDock(QWidget):
    """Tree view of project structure — resources grouped by kind, data sources, results."""

    resourceSelected = Signal(dict)

    def __init__(self, parent=None):
        super().__init__(parent)
        self._project: dict | None = None
        self._project_path: str | None = None
        self.tree = QTreeWidget()
        self.tree.setHeaderHidden(True)
        self.tree.setContextMenuPolicy(Qt.ContextMenuPolicy.CustomContextMenu)
        self.tree.customContextMenuRequested.connect(self._context_menu)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(8, 8, 8, 8)
        layout.addWidget(self.tree)
        self.set_empty_state()

    def set_empty_state(self) -> None:
        self.tree.clear()
        root = QTreeWidgetItem([tr("project.none")])
        root.setDisabled(True)
        self.tree.addTopLevelItem(root)

    def set_project(self, project: dict | None, project_path: str | None = None) -> None:
        self._project = project
        self._project_path = project_path
        self.tree.clear()
        if not project:
            self.set_empty_state()
            return

        project_name = project.get("project_name") or tr("project.untitled")
        root = QTreeWidgetItem([project_name])
        root.setExpanded(True)
        self.tree.addTopLevelItem(root)

        if project_path:
            location = QTreeWidgetItem([f"{tr('project.file')}: {Path(project_path).name}"])
            location.setToolTip(0, str(project_path))
            root.addChild(location)

        # Group resources by kind
        resources = project.get("resources", [])
        grouped = self._group_by_kind(resources)
        for kind, items in grouped.items():
            self._add_resource_group(root, f"{kind} ({len(items)})", items)

        self._add_resource_group(root, tr("project.data_sources"), project.get("data_sources", []))
        self._add_resource_group(root, tr("project.results"), project.get("result_history", []))
        self.tree.expandAll()

    def _group_by_kind(self, resources: list[dict]) -> dict[str, list[dict]]:
        groups: dict[str, list[dict]] = {}
        for r in resources:
            kind = r.get("source_type") or r.get("kind") or tr("resource.kind.other")
            label = tr(f"resource.kind.{kind}")
            groups.setdefault(label, []).append(r)
        return groups

    def _add_resource_group(self, parent: QTreeWidgetItem, title: str,
                            items: Iterable[dict]) -> None:
        group = QTreeWidgetItem([title])
        parent.addChild(group)
        count = 0
        for item in items or []:
            count += 1
            label = (
                item.get("name")
                or item.get("title")
                or item.get("source_path")
                or item.get("path")
                or tr("project.item", count=count)
            )
            child = QTreeWidgetItem([str(label)])
            path_val = item.get("source_path") or item.get("path")
            if path_val:
                child.setToolTip(0, str(path_val))
            child.setData(0, 256, item)
            group.addChild(child)
        if count == 0:
            empty = QTreeWidgetItem([tr("project.empty")])
            empty.setDisabled(True)
            group.addChild(empty)

    def _context_menu(self, pos) -> None:
        item = self.tree.itemAt(pos)
        if not item:
            return
        data = item.data(0, 256)
        if not data:
            return
        menu = QMenu(self)
        remove_action = menu.addAction(tr("action.remove_resource"))
        remove_action.triggered.connect(lambda: self._remove_resource(data))
        menu.exec_(self.tree.viewport().mapToGlobal(pos))

    def _remove_resource(self, resource: dict) -> None:
        rid = resource.get("resource_id") or resource.get("id")
        if rid:
            window = self.window()
            if hasattr(window, "_ctx"):
                window._ctx.project_service.remove_resource(rid)
                window._refresh_project_views()

    def retranslate_ui(self) -> None:
        self.set_project(self._project, self._project_path)

File: migration_project/cpp_qt/src/docks/ProjectDock.cpp
#include "ProjectDock.h"
#include "../I18n.h"

#include <QVBoxLayout>
#include <QJsonArray>
#include <QFileInfo>

ProjectDock::ProjectDock(QWidget* parent)
    : QWidget(parent)
{
    m_tree = new QTreeWidget(this);
    m_tree->setHeaderHidden(true);

    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(8, 8, 8, 8);
    layout->addWidget(m_tree);

    setEmptyState();
}

void ProjectDock::setEmptyState() {
    m_tree->clear();
    auto* root = new QTreeWidgetItem({I18n::instance().tr("project.none")});
    root->setFlags(root->flags() & ~Qt::ItemIsEnabled);
    m_tree->addTopLevelItem(root);
}

void ProjectDock::setProject(const QJsonObject& project, const QString& projectPath) {
    m_project = project;
    m_projectPath = projectPath;
    m_tree->clear();

    if (project.isEmpty()) {
        setEmptyState();
        return;
    }

    QString name = project.value("project_name").toString();
    if (name.isEmpty())
        name = I18n::instance().tr("project.untitled");

    auto* root = new QTreeWidgetItem({name});
    root->setExpanded(true);
    m_tree->addTopLevelItem(root);

    if (!projectPath.isEmpty()) {
        auto* loc = new QTreeWidgetItem({
            QString("%1: %2").arg(I18n::instance().tr("project.file"),
                                  QFileInfo(projectPath).fileName())
        });
        loc->setToolTip(0, projectPath);
        root->addChild(loc);
    }

    addResourceGroup(root, I18n::instance().tr("project.resources"), project.value("resources").toArray());
    addResourceGroup(root, I18n::instance().tr("project.data_sources"), project.value("data_sources").toArray());
    addResourceGroup(root, I18n::instance().tr("project.results"), project.value("result_history").toArray());

    m_tree->expandAll();
}

void ProjectDock::addResourceGroup(QTreeWidgetItem* parent, const QString& title, const QJsonArray& items) {
    auto* group = new QTreeWidgetItem({title});
    parent->addChild(group);

    if (items.isEmpty()) {
        auto* empty = new QTreeWidgetItem({I18n::instance().tr("project.empty")});
        empty->setFlags(empty->flags() & ~Qt::ItemIsEnabled);
        group->addChild(empty);
        return;
    }

    int count = 0;
    for (const auto& val : items) {
        QJsonObject item = val.toObject();
        ++count;
        QString label = item.value("name").toString()
            .isEmpty() ? item.value("title").toString() : item.value("name").toString();
        QString path = item.value("source_path").toString();
        if (path.isEmpty()) path = item.value("path").toString();
        if (label.isEmpty())
            label = path.isEmpty() ? I18n::instance().tr("project.item").arg(count) : QFileInfo(path).fileName();

        auto* child = new QTreeWidgetItem({label});
        if (!path.isEmpty())
            child->setToolTip(0, path);
        group->addChild(child);
    }
}

void ProjectDock::retranslateUi() {
    setProject(m_project, m_projectPath);
}
